keps/sig-auth/4872-harden-kubelet-cert-validation/README.md

```diff
@@ -0,0 +1,389 @@
+* Prevent rogue nodes from using certificates issued for IPs they do not own.
+
+### Non-Goals
+
+* This proposal does not address certificate validation for clients other than the Kube API server, such as metrics scrapers. However, we'll consider an implementation in client-go that could be used by those other clients.
+
+## Proposal
+
+We propose that the Kube API server is modified to validate the Common Name (CN) of the kubelet's serving certificate is equal to `system:node:<nodename>`.
+`nodename` is the name of the Node object as reported by the kubelet. When the Kube-API server connects to the kubelet server (e.g. for logs, exec, port-forward), it always knows the Node it's connecting to.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster administrator, I want to ensure that kubelet serving certificates are validated based on the node name, reducing the risk of IP-based impersonation attacks.
+
+#### Story 2
+
+As a cluster administrator using custom serving certificates for the kubelet server, I want to be able to disable the Subject's CN validation.
+
+### Notes/Constraints/Caveats (Optional)
+
+When the kubelet requests a certificate through a CSR, it sets the CN to `system:node:<nodename>`, enforced by the admission controller as per [PR \#126015](https://github.com/kubernetes/kubernetes/pull/126015).
+
+However, certificates issued manually or through other mechanisms may not follow this convention.
+With the new validation, any certificate not following this `system:node:<nodename>` convention will be deemed invalid by the Kube API server.
+This will require cluster administrators to reissue any non-conforming certificates before enabling this feature.
+
+### Risks and Mitigations
+
+This could disrupt existing clusters that are using custom kubelet serving certificates.
+These clusters will need to reissue their certificates before enabling this feature. We will allow to disable the validation through a command-line flag to allow for a smooth transition.
+
+## Design Details
+
+### Enabling the feature
+
+We will introduce a feature flag `KubeletCertCNValidation` that will gate the usage of the new validation.
+This gate will start off by default in Alpha, will be turned on by default in Beta and will be removed in GA.
+
+In addition, we will allow to disable the validation through a command-line flag `--disable-kubelet-cert-cn-validation`.
+This flag can only be set if the `KubeletCertCNValidation` feature flag is enabled.
+This flag will allow cluster administrators to opt-out of this validation if they are using custom kubelet serving certificates that don't follow the `system:node:<nodename>` convention even after the feature gate is removed.
+
+#### Metrics
+
+In order to help cluster administrators determine if it's safe to enable the feature, we propose to add a new metric `kube_apiserver_validation_kubelet_cert_cn_errors` that will track the number of errors due to the new CN validation.
```

Comment:
I would call the metric something like `kube_apiserver_validation_kubelet_cert_cn_total` and have a `success` and `failure` label.  Then we would increment either counter as needed.